src/compiler/checker.ts

```diff
@@ -25163,7 +25163,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
      * variable T[P] (i.e. we treat the type T[P] as the type variable we're inferring for).
      */
     function inferTypeForHomomorphicMappedType(source: Type, target: MappedType, constraint: IndexType): Type | undefined {
-        const cacheKey = source.id + "," + target.id + "," + constraint.id;
+        const cacheKey = "H" + source.id + "," + target.id + "," + constraint.id;

```

Comment:
The same keys (and results) were shared by `inferTypeForHomomorphicMappedType` and `inferReverseMappedType` and that caused a regression. This means that each of those could get a separate cache as they are not meant to be shared anyway. I didn't bother doing that here but if the team thinks that's a good idea I could do it.

Reply:
I would recommend using a different cache, since there's no reason for them to be the same.

Reply:
done